File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashViewModel.java
package com.github.dedis.popstellar.ui.digitalcash;

import android.app.Application;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.*;

import com.github.dedis.popstellar.SingleEvent;
import com.github.dedis.popstellar.model.network.method.message.MessageGeneral;
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.*;
import com.github.dedis.popstellar.model.objects.Channel;
import com.github.dedis.popstellar.model.objects.RollCall;
import com.github.dedis.popstellar.model.objects.digitalcash.TransactionObject;
import com.github.dedis.popstellar.model.objects.security.*;
import com.github.dedis.popstellar.model.objects.view.LaoView;
import com.github.dedis.popstellar.repository.*;
import com.github.dedis.popstellar.repository.remote.GlobalNetworkManager;
import com.github.dedis.popstellar.utility.error.UnknownLaoException;
import com.github.dedis.popstellar.utility.error.keys.KeyException;
import com.github.dedis.popstellar.utility.error.keys.NoRollCallException;
import com.github.dedis.popstellar.utility.security.KeyManager;
import com.google.gson.Gson;

import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.util.*;
import java.util.stream.Collectors;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import io.reactivex.Observable;
import io.reactivex.*;

@HiltViewModel
public class DigitalCashViewModel extends AndroidViewModel {

  public static final String TAG = DigitalCashViewModel.class.getSimpleName();

  private String laoId;
  private static final String LAO_FAILURE_MESSAGE = "failed to retrieve lao";
  private static final String RECEIVER_KEY_ERROR = "Error on the receiver s public key";
  private static final String COIN = "coin";

  private static final String TYPE = "P2PKH";
  private static final int VERSION = 1;

  public static final int NOTHING_SELECTED = -1;
  public static final int MIN_LAO_COIN = 0;

  /*
   * LiveData objects for capturing events
   */
  private final MutableLiveData<SingleEvent<Boolean>> postTransactionEvent =
      new MutableLiveData<>();

  /* Update the receipt after sending a transaction */
  private final MutableLiveData<SingleEvent<String>> updateReceiptAddressEvent =
      new MutableLiveData<>();
  private final MutableLiveData<SingleEvent<String>> updateReceiptAmountEvent =
      new MutableLiveData<>();


  /*
   * Dependencies for this class
   */
  private final LAORepository laoRepository;
  private final RollCallRepository rollCallRepo;
  private final DigitalCashRepository digitalCashRepo;
  private final GlobalNetworkManager networkManager;
  private final Gson gson;
  private final KeyManager keyManager;

  @Inject
  public DigitalCashViewModel(
      @NonNull Application application,
      LAORepository laoRepository,
      RollCallRepository rollCallRepo,
      DigitalCashRepository digitalCashRepo,
      GlobalNetworkManager networkManager,
      Gson gson,
      KeyManager keyManager) {
    super(application);
    this.laoRepository = laoRepository;
    this.rollCallRepo = rollCallRepo;
    this.digitalCashRepo = digitalCashRepo;
    this.networkManager = networkManager;
    this.gson = gson;
    this.keyManager = keyManager;
  }

  public void setLaoId(String laoId) {
    this.laoId = laoId;
  }

  public LiveData<SingleEvent<Boolean>> getPostTransactionEvent() {
    return postTransactionEvent;
  }

  public void postTransactionEvent() {
    postTransactionEvent.postValue(new SingleEvent<>(true));
  }

  public LiveData<SingleEvent<String>> getUpdateReceiptAddressEvent() {
    return updateReceiptAddressEvent;
  }

  public void updateReceiptAddressEvent(String address) {
    updateReceiptAddressEvent.postValue(new SingleEvent<>(address));
  }

  public LiveData<SingleEvent<String>> getUpdateReceiptAmountEvent() {
    return updateReceiptAmountEvent;
  }

  public void updateReceiptAmountEvent(String amount) {
    updateReceiptAmountEvent.postValue(new SingleEvent<>(amount));
  }

  public void requireToPutAnAmount() {
    Toast.makeText(
            getApplication().getApplicationContext(),
            "Please enter a positive amount of LAOcoin",
            Toast.LENGTH_LONG)
        .show();
  }

  public void requireToPutLAOMember() {
    Toast.makeText(
            getApplication().getApplicationContext(),
            "Please select a LAOMember",
            Toast.LENGTH_LONG)
        .show();
  }

  /*
   * Methods that modify the state or post an Event to update the UI.
   */
  public PublicKey getPublicKeyOutString(String encodedPub) throws NoRollCallException {
    for (PublicKey current : Objects.requireNonNull(getAttendeesFromLastRollCall())) {
      if (current.getEncoded().equals(encodedPub)) {
        return current;
      }
    }
    return null;
  }

  private long computeOutputs(Map.Entry<String, String> current, List<Output> outputs) {
    try {
      PublicKey pub = getPublicKeyOutString(current.getKey());
      long amount = Long.parseLong(current.getValue());
      Output addOutput = new Output(amount, new ScriptOutput(TYPE, pub.computeHash()));
      outputs.add(addOutput);
      return amount;
    } catch (Exception e) {
      Log.e(TAG, RECEIVER_KEY_ERROR, e);
      return 0;
    }
  }

  /**
   * Post a transaction to your channel
   *
   * <p>Publish a Message General containing a PostTransaction data
   *
   * @return a Single emitting the sent transaction object
   */
  public Completable postTransaction(
      Map<String, String> receiverValues, long lockTime, boolean coinBase) {

    /* Check if a Lao exist */
    LaoView laoView;
    try {
      laoView = getLao();
    } catch (UnknownLaoException e) {
      Log.e(TAG, LAO_FAILURE_MESSAGE);
      return Completable.error(new UnknownLaoException());
    }

    // Find correct keypair
    return Single.fromCallable(() -> coinBase ? keyManager.getMainKeyPair() : this.getValidToken())
        .flatMapCompletable(
            keyPair -> {
              PostTransactionCoin postTxn =
                  createPostTransaction(keyPair, receiverValues, lockTime, coinBase);
              MessageGeneral msg = new MessageGeneral(keyPair, postTxn, gson);
              Channel channel = laoView.getChannel().subChannel(COIN);
              return networkManager
                  .getMessageSender()
                  .publish(channel, msg)
                  .doOnComplete(
                      () -> Log.d(TAG, "Successfully sent post transaction message : " + postTxn));
            });
  }

  private PostTransactionCoin createPostTransaction(
      KeyPair keyPair, Map<String, String> receiverValues, long lockTime, boolean coinBase)
      throws GeneralSecurityException {
    // first make the output
    List<Output> outputs = new ArrayList<>();
    long amountFromReceiver = 0;
    for (Map.Entry<String, String> current : receiverValues.entrySet()) {
      amountFromReceiver += computeOutputs(current, outputs);
    }

    // Then make the inputs
    // First there would be only one Input

    // Case no transaction before
    String transactionHash = TransactionObject.TX_OUT_HASH_COINBASE;
    int index = 0;

    List<Input> inputs = new ArrayList<>();
    List<TransactionObject> transactions = getTransactionsForUser(keyPair.getPublicKey());
    if (transactions != null && !coinBase) {
      processNotCoinbaseTransaction(keyPair, outputs, amountFromReceiver, inputs);
    } else {
      inputs.add(
          processSignInput(
              keyPair, outputs, Collections.singletonMap(transactionHash, index), transactionHash));
    }

    Transaction transaction = new Transaction(VERSION, inputs, outputs, lockTime);

    return new PostTransactionCoin(transaction);
  }

  private Input processSignInput(
      KeyPair keyPair,
      List<Output> outputs,
      Map<String, Integer> transactionInpMap,
      String currentHash)
      throws GeneralSecurityException {
    Signature sig =
        keyPair.sign(
            new Base64URLData(
                Transaction.computeSigOutputsPairTxOutHashAndIndex(outputs, transactionInpMap)
                    .getBytes(StandardCharsets.UTF_8)));
    return new Input(
        currentHash,
        transactionInpMap.get(currentHash),
        new ScriptInput(TYPE, keyPair.getPublicKey(), sig));
  }

  public LAORepository getLaoRepository() {
    return laoRepository;
  }

  public PublicKey getOwnKey() {
    return keyManager.getMainPublicKey();
  }

  @Nullable
  public Set<PublicKey> getAttendeesFromLastRollCall() throws NoRollCallException {
    return rollCallRepo.getLastClosedRollCall(laoId).getAttendees();
  }

  @Nullable
  public PublicKey getOrganizer() throws UnknownLaoException {
    return getLao().getOrganizer();
  }

  @Nullable
  public List<String> getAttendeesFromTheRollCallList() throws NoRollCallException {
    return Objects.requireNonNull(getAttendeesFromLastRollCall()).stream()
        .map(Base64URLData::getEncoded)
        .collect(Collectors.toList());
  }

  public LaoView getLao() throws UnknownLaoException {
    return laoRepository.getLaoView(laoId);
  }

  public PoPToken getValidToken() throws KeyException {
    return keyManager.getValidPoPToken(laoId, rollCallRepo.getLastClosedRollCall(laoId));
  }

  public boolean canPerformTransaction(
      String currentAmount, String currentPublicKeySelected, int radioGroup) {
    if ((currentAmount.isEmpty()) || (Integer.parseInt(currentAmount) < MIN_LAO_COIN)) {
      // create in View Model a function that toast : please enter amount
      requireToPutAnAmount();
      return false;
    } else if (currentPublicKeySelected.isEmpty() && (radioGroup == NOTHING_SELECTED)) {
      // create in View Model a function that toast : please enter key
      requireToPutLAOMember();
      return false;
    } else {
      return true;
    }
  }

  private void processNotCoinbaseTransaction(
      KeyPair keyPair, List<Output> outputs, long amountFromReceiver, List<Input> inputs)
      throws GeneralSecurityException {
    int index;
    String transactionHash;
    List<TransactionObject> transactions = getTransactionsForUser(keyPair.getPublicKey());

    long amountSender = getUserBalance(keyPair.getPublicKey()) - amountFromReceiver;
    Output outputSender =
        new Output(amountSender, new ScriptOutput(TYPE, keyPair.getPublicKey().computeHash()));
    outputs.add(outputSender);
    Map<String, Integer> transactionInpMap = new HashMap<>();
    for (TransactionObject transactionPrevious : transactions) {
      transactionHash = transactionPrevious.getTransactionId();
      index = transactionPrevious.getIndexTransaction(keyPair.getPublicKey());
      transactionInpMap.put(transactionHash, index);
    }

    for (String currentHash : transactionInpMap.keySet()) {
      inputs.add(processSignInput(keyPair, outputs, transactionInpMap, currentHash));
    }
  }

  public List<TransactionObject> getTransactionsForUser(PublicKey user) {
    return digitalCashRepo.getTransactions(laoId, user);
  }

  public Observable<List<TransactionObject>> getTransactionsObservable() {
    try {
      return digitalCashRepo.getTransactionsObservable(laoId, getValidToken().getPublicKey());
    } catch (KeyException e) {
      return Observable.error(e);
    }
  }

  public Observable<Set<RollCall>> getRollCallsObservable() {
    return rollCallRepo.getRollCallsObservableInLao(laoId);
  }

  public long getUserBalance(PublicKey user) {
    return digitalCashRepo.getUserBalance(laoId, user);
  }

  public long getOwnBalance() throws KeyException {
    return getUserBalance(getValidToken().getPublicKey());
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/digitalcash/DigitalCashViewModel.kt
package com.github.dedis.popstellar.ui.lao.digitalcash

import android.app.Application
import android.widget.Toast
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.github.dedis.popstellar.R
import com.github.dedis.popstellar.SingleEvent
import com.github.dedis.popstellar.model.network.method.message.MessageGeneral
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.Input
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.Output
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.PostTransactionCoin
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.ScriptInput
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.ScriptOutput
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.Transaction
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.Transaction.Companion.computeSigOutputsPairTxOutHashAndIndex
import com.github.dedis.popstellar.model.objects.RollCall
import com.github.dedis.popstellar.model.objects.digitalcash.TransactionObject
import com.github.dedis.popstellar.model.objects.security.Base64URLData
import com.github.dedis.popstellar.model.objects.security.KeyPair
import com.github.dedis.popstellar.model.objects.security.PoPToken
import com.github.dedis.popstellar.model.objects.security.PublicKey
import com.github.dedis.popstellar.model.objects.view.LaoView
import com.github.dedis.popstellar.repository.DigitalCashRepository
import com.github.dedis.popstellar.repository.LAORepository
import com.github.dedis.popstellar.repository.RollCallRepository
import com.github.dedis.popstellar.repository.WitnessingRepository
import com.github.dedis.popstellar.repository.remote.GlobalNetworkManager
import com.github.dedis.popstellar.utility.error.ErrorUtils.logAndShow
import com.github.dedis.popstellar.utility.error.UnknownLaoException
import com.github.dedis.popstellar.utility.error.keys.KeyException
import com.github.dedis.popstellar.utility.error.keys.NoRollCallException
import com.github.dedis.popstellar.utility.security.KeyManager
import com.google.gson.Gson
import dagger.hilt.android.lifecycle.HiltViewModel
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import java.nio.charset.StandardCharsets
import java.security.GeneralSecurityException
import java.util.Collections
import java.util.stream.Collectors
import javax.inject.Inject
import timber.log.Timber

@HiltViewModel
class DigitalCashViewModel
@Inject
constructor(
    application: Application,
    /*
     * Dependencies for this class
     */
    val laoRepository: LAORepository,
    private val rollCallRepo: RollCallRepository,
    private val digitalCashRepo: DigitalCashRepository,
    private val witnessingRepo: WitnessingRepository,
    private val networkManager: GlobalNetworkManager,
    private val gson: Gson,
    private val keyManager: KeyManager
) : AndroidViewModel(application) {
  var laoId: String? = null

  /*
   * LiveData objects for capturing events
   */
  private val postTransactionEvent = MutableLiveData<SingleEvent<Boolean>>()

  /* Update the receipt after sending a transaction */
  private val updateReceiptAddressEvent = MutableLiveData<SingleEvent<String>>()
  private val updateReceiptAmountEvent = MutableLiveData<SingleEvent<String>>()

  fun getPostTransactionEvent(): LiveData<SingleEvent<Boolean>> {
    return postTransactionEvent
  }

  fun postTransactionEvent() {
    postTransactionEvent.postValue(SingleEvent(true))
  }

  fun getUpdateReceiptAddressEvent(): LiveData<SingleEvent<String>> {
    return updateReceiptAddressEvent
  }

  fun updateReceiptAddressEvent(address: String?) {
    updateReceiptAddressEvent.postValue(SingleEvent(address))
  }

  fun getUpdateReceiptAmountEvent(): LiveData<SingleEvent<String>> {
    return updateReceiptAmountEvent
  }

  fun updateReceiptAmountEvent(amount: String?) {
    updateReceiptAmountEvent.postValue(SingleEvent(amount))
  }

  private fun requireToPutAnAmount() {
    Toast.makeText(
            getApplication<Application>().applicationContext,
            R.string.digital_cash_amount_min_indication,
            Toast.LENGTH_LONG)
        .show()
  }

  private fun requireToPutLAOMember() {
    Toast.makeText(
            getApplication<Application>().applicationContext,
            R.string.digital_cash_select_lao_member_indication,
            Toast.LENGTH_LONG)
        .show()
  }

  /*
   * Methods that modify the state or post an Event to update the UI.
   */
  @Throws(NoRollCallException::class)
  fun getPublicKeyOutString(encodedPub: String): PublicKey? {
    for (current in attendeesFromLastRollCall) {
      if (current.encoded == encodedPub) {
        return current
      }
    }
    return null
  }

  private fun computeOutputs(
      current: Map.Entry<String, String>,
      outputs: MutableList<Output>
  ): Long {
    return try {
      val pub = getPublicKeyOutString(current.key)
      val amount = current.value.toLong()
      val addOutput = Output(amount, ScriptOutput(TYPE, pub!!.computeHash()))
      outputs.add(addOutput)

      amount
    } catch (e: Exception) {
      Timber.tag(TAG).e(e, RECEIVER_KEY_ERROR)
      0
    }
  }

  /**
   * Post a transaction to your channel
   *
   * Publish a Message General containing a PostTransaction data
   *
   * @return a Single emitting the sent transaction object
   */
  fun postTransaction(
      receiverValues: Map<String, String>,
      lockTime: Long,
      coinBase: Boolean
  ): Completable {

    /* Check if a Lao exist */
    val laoView: LaoView =
        try {
          lao
        } catch (e: UnknownLaoException) {
          Timber.tag(TAG).e(e, LAO_FAILURE_MESSAGE)
          return Completable.error(UnknownLaoException())
        }

    // Find correct keypair
    return Single.fromCallable { if (coinBase) keyManager.mainKeyPair else validToken }
        .flatMapCompletable { keyPair: KeyPair ->
          val postTxn = createPostTransaction(keyPair, receiverValues, lockTime, coinBase)
          val msg = MessageGeneral(keyPair, postTxn, gson)
          val channel = laoView.channel.subChannel(COIN)

          networkManager.messageSender.publish(channel, msg).doOnComplete {
            Timber.tag(TAG).d("Successfully sent post transaction message : %s", postTxn)
          }
        }
  }

  @Throws(GeneralSecurityException::class)
  private fun createPostTransaction(
      keyPair: KeyPair,
      receiverValues: Map<String, String>,
      lockTime: Long,
      coinBase: Boolean
  ): PostTransactionCoin {
    // first make the output
    val outputs: MutableList<Output> = ArrayList()
    var amountFromReceiver: Long = 0
    for (current in receiverValues.entries) {
      amountFromReceiver += computeOutputs(current, outputs)
    }

    // Then make the inputs
    // First there would be only one Input

    // Case no transaction before
    val transactionHash = TransactionObject.TX_OUT_HASH_COINBASE
    val index = 0
    val inputs: MutableList<Input> = ArrayList()
    val transactions = getTransactionsForUser(keyPair.publicKey)
    if (transactions != null && !coinBase) {
      processNotCoinbaseTransaction(keyPair, outputs, amountFromReceiver, inputs)
    } else {
      inputs.add(
          processSignInput(
              keyPair, outputs, Collections.singletonMap(transactionHash, index), transactionHash))
    }
    val transaction = Transaction(VERSION, inputs, outputs, lockTime)

    return PostTransactionCoin(transaction)
  }

  @Throws(GeneralSecurityException::class)
  private fun processSignInput(
      keyPair: KeyPair,
      outputs: List<Output>,
      transactionInpMap: Map<String, Int>,
      currentHash: String
  ): Input {
    val sig =
        keyPair.sign(
            Base64URLData(
                computeSigOutputsPairTxOutHashAndIndex(outputs, transactionInpMap)
                    .toByteArray(StandardCharsets.UTF_8)))

    return Input(
        currentHash, transactionInpMap[currentHash]!!, ScriptInput(TYPE, keyPair.publicKey, sig))
  }

  @get:Throws(NoRollCallException::class)
  val attendeesFromLastRollCall: Set<PublicKey>
    get() = rollCallRepo.getLastClosedRollCall(laoId!!).attendees

  @get:Throws(UnknownLaoException::class)
  val organizer: PublicKey
    get() = lao.organizer

  @get:Throws(NoRollCallException::class)
  val attendeesFromTheRollCallList: List<String>
    get() =
        attendeesFromLastRollCall.stream().map(Base64URLData::encoded).collect(Collectors.toList())

  @get:Throws(UnknownLaoException::class)
  val lao: LaoView
    get() = laoRepository.getLaoView(laoId!!)

  val witnesses: Set<PublicKey>
    get() = witnessingRepo.getWitnesses(laoId!!)

  @get:Throws(KeyException::class)
  val validToken: PoPToken
    get() = keyManager.getValidPoPToken(laoId!!, rollCallRepo.getLastClosedRollCall(laoId!!))

  val ownKey: PublicKey
    get() = keyManager.mainPublicKey

  fun canPerformTransaction(
      currentAmount: String,
      currentPublicKeySelected: String,
      radioGroup: Int
  ): Boolean {
    if (currentAmount.isEmpty()) {
      // ask the user to fill the amount box
      requireToPutAnAmount()
    }

    val parsedAmount =
        try {
          currentAmount.toInt()
        } catch (e: NumberFormatException) {
          // Overflow in the amount (no characters or negative numbers can be inserted)
          logAndShow(
              getApplication<Application>().applicationContext,
              TAG,
              R.string.digital_cash_amount_inserted_error)
          return false
        }

    if (parsedAmount <= MIN_LAO_COIN) {
      logAndShow(
          getApplication<Application>().applicationContext,
          TAG,
          R.string.digital_cash_amount_min_indication)
      return false
    } else if (currentPublicKeySelected.isEmpty() && radioGroup == NOTHING_SELECTED) {
      // create in View Model a function that toast : please enter key
      requireToPutLAOMember()
      return false
    }

    return true
  }

  @Throws(GeneralSecurityException::class)
  private fun processNotCoinbaseTransaction(
      keyPair: KeyPair,
      outputs: MutableList<Output>,
      amountFromReceiver: Long,
      inputs: MutableList<Input>
  ) {
    var index: Int
    var transactionHash: String
    val transactions = getTransactionsForUser(keyPair.publicKey)!!

    val amountSender = getUserBalance(keyPair.publicKey) - amountFromReceiver
    val outputSender = Output(amountSender, ScriptOutput(TYPE, keyPair.publicKey.computeHash()))
    outputs.add(outputSender)

    val transactionInpMap: MutableMap<String, Int> = HashMap()
    for (transactionPrevious in transactions) {
      transactionHash = transactionPrevious.transactionId
      index = transactionPrevious.getIndexTransaction(keyPair.publicKey)
      transactionInpMap[transactionHash] = index
    }

    for (currentHash in transactionInpMap.keys) {
      inputs.add(processSignInput(keyPair, outputs, transactionInpMap, currentHash))
    }
  }

  private fun getTransactionsForUser(user: PublicKey): List<TransactionObject>? {
    return digitalCashRepo.getTransactions(laoId!!, user)
  }

  val transactionsObservable: Observable<List<TransactionObject>>
    get() =
        try {
          digitalCashRepo.getTransactionsObservable(laoId!!, validToken.publicKey)
        } catch (e: KeyException) {
          Observable.error(e)
        }

  val rollCallsObservable: Observable<Set<RollCall>>
    get() = rollCallRepo.getRollCallsObservableInLao(laoId!!)

  fun getUserBalance(user: PublicKey): Long {
    return digitalCashRepo.getUserBalance(laoId!!, user)
  }

  @get:Throws(KeyException::class)
  val ownBalance: Long
    get() = getUserBalance(validToken.publicKey)

  companion object {
    val TAG: String = DigitalCashViewModel::class.java.simpleName
    private const val LAO_FAILURE_MESSAGE = "failed to retrieve lao"
    private const val RECEIVER_KEY_ERROR = "Error on the receiver s public key"
    private const val COIN = "coin"
    private const val TYPE = "P2PKH"
    private const val VERSION = 1
    const val NOTHING_SELECTED = -1
    const val MIN_LAO_COIN = 0
  }
}
